### PR TITLE
[16.0] Add sale_cart_step + shopinvader_api_cart_step

### PR DESCRIPTION
**Pull request by simahawk**
Addresses https://github.com/shopinvader/odoo-shopinvader/issues/1511

Depends on:
- [x] https://github.com/shopinvader/odoo-shopinvader/pull/1534 

TODO: what to do w/ old shopinvader modules?
I'm not sure this is needed.
If you don't use sale_cart_step you won't need any migration.
TBH I'd like not to waste time on migrating `shopinvader` and/or `shopinvader_restapi`.

### CODE REVIEW

File: shopinvader_api_cart_step/schemas/cart.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class CartUpdateInput(BaseCartUpdateInput, extends=True):
+
+    current_step: str | None = None
+    next_step: str | None = None
+
+    def convert_to_sale_write(self):
+        vals = super().convert_to_sale_write()
+        # NOTE: is not possible to use `env` to retrive data here
```
Comment (simahawk) [inline shopinvader_api_cart_step/schemas/cart.py]: IMO as done in other converters we should probably pass the current SO record to this method and avoid to override `_update_prepare_data` to retrieve data that relies on ORM.

**Review by mmequignon — APPROVED** [review]

File: shopinvader_api_cart_step/schemas/cart.py (hunk 1 above)
Comment (simahawk) [reply shopinvader_api_cart_step/schemas/cart.py]: In general I tend to say that these methods should get the env at least...
@lmignon WDYT?

Comment (lmignon) [reply shopinvader_api_cart_step/schemas/cart.py]: > IMO as done in other converters we should probably pass the current SO record to this method and avoid to override `_update_prepare_data` to retrieve data that relies on ORM.

:+1:

Comment (lmignon) [reply shopinvader_api_cart_step/schemas/cart.py]: @sebastienbeau @marielejeune @qgroulard What do you think about this change, and is it likely to have an impact on your current devs?

Comment (marielejeune) [reply shopinvader_api_cart_step/schemas/cart.py]: It seems OK for me

**Review by cyrilmanuel — APPROVED** [review]

**Comment by cyrilmanuel** [comment]
hi @lmignon , can you merge this PR ?

File: shopinvader_api_cart/routers/cart.py [hunk 2]
```diff
@@ -236,5 +236,8 @@ def _update(self, partner, data, uuid):
         cart = self.env["sale.order"]._find_open_cart(partner.id, uuid)
         if not cart:
             cart = self.env["sale.order"]._create_empty_cart(partner.id)
-        cart.write(data.convert_to_sale_write())
+        cart.write(self._update_prepare_data(data))
         return cart
+
+    def _update_prepare_data(self, data):
+        return data.convert_to_sale_write()
```
Comment (lmignon) [inline shopinvader_api_cart/routers/cart.py]: IMO we should modify the `convert_to_sale_write` method to pass the environment and avoid to scattering the logic to convert the model into a dict at different places.

**Review by lmignon — CHANGES_REQUESTED** [review]

File: shopinvader_api_cart/routers/cart.py (hunk 2 above)
Comment (simahawk) [reply shopinvader_api_cart/routers/cart.py]: I agree... can I introduce a breaking change then? :grin:

Comment (lmignon) [reply shopinvader_api_cart/routers/cart.py]: Yes

Comment (simahawk) [reply shopinvader_api_cart/routers/cart.py]: I'm passing the `cart` now.
In general I think it makes sense to rework existing methods like this:
* pass and env for non write methods
* pass the real record for write methods

Comment (simahawk) [reply shopinvader_api_cart/routers/cart.py]: I split this change to https://github.com/shopinvader/odoo-shopinvader/pull/1534
@lmignon could we fast track that?

**Review by lmignon — APPROVED** [review]

**Comment by lmignon** [comment]
#1534 is merged....

**Pull request merged**